The Windows build seems to be failing (I already tried restarting it to see if it was a flake):

```
[ERROR] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 2.447 s <<< FAILURE! - in io.quarkus.mailer.InjectionTest
[ERROR] testInjection  Time elapsed: 0.063 s  <<< ERROR!
java.lang.NullPointerException
	at io.quarkus.qute.runtime.VariantTemplateProducer$VariantTemplateInstanceImpl.<init>(VariantTemplateProducer.java:131)
	at io.quarkus.qute.runtime.VariantTemplateProducer$VariantTemplateImpl.instance(VariantTemplateProducer.java:105)
	at io.quarkus.mailer.runtime.MailTemplateProducer$2.instance(MailTemplateProducer.java:75)

```

--- a/docs/src/main/asciidoc/mailer.adoc
+++ b/docs/src/main/asciidoc/mailer.adoc
@@ -197,6 +197,33 @@ By spec, when you create the inline attachment, the content-id must be structure
 If you don't wrap your content-id between `<>`, it is automatically wrapped for you.
 When you want to reference your attachment, for instance in the `src` attribute, use `cid:id@domain` (without the `<` and `>`).
 
+== Message Body Based on Qute Templates
+
+It's also possible to inject a mail template, where the message body is created automatically using link:qute[Qute templates].
+
+[source, java]
+----
+@Inject
+MailTemplate hello; <1>
+
+@GET
+@Path("/mail")
+public CompletionStage<Response> send() {
+    return hello.to("to@acme.org") <2>
+       .subject("Hello from Qute template")
+       // the template looks like: Hello {name}!
+       .data("name", "John") <3>
+       .send() <4>
+       .thenApply(x -> Response.accepted().build());
+}
+----
+<1> If there is no `@ResourcePath` qualifier provided, the field name is used to locate the template. In this particular case, we will use the `hello.html` and `hello.txt` templates to create the message body.
+<2> Create a mail template instance and set the recipient.
+<3> Set the data used in the template.
+<4> `MailTemplate.send()` triggers the rendering and, once finished, sends the e-mail via a `Mailer` instance. 
+
+TIP: Injected mail templates are validated during build. If there is no matching template in `src/main/resources/templates` the build fails.
+
 == Testing email sending
 
 Because it is very inconvenient to send emails during development and testing, you can set the `quarkus.mailer.mock` boolean

--- a/extensions/mailer/deployment/pom.xml
+++ b/extensions/mailer/deployment/pom.xml
@@ -22,6 +22,10 @@
         </dependency>
         <dependency>
             <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-qute-deployment</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>io.quarkus</groupId>
             <artifactId>quarkus-mailer</artifactId>
         </dependency>
 

--- a/extensions/mailer/deployment/src/main/java/io/quarkus/mailer/deployment/MailerProcessor.java
+++ b/extensions/mailer/deployment/src/main/java/io/quarkus/mailer/deployment/MailerProcessor.java
@@ -1,7 +1,19 @@
 package io.quarkus.mailer.deployment;
 
+import java.io.File;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+import org.jboss.jandex.AnnotationInstance;
+import org.jboss.jandex.DotName;
+
 import io.quarkus.arc.deployment.AdditionalBeanBuildItem;
 import io.quarkus.arc.deployment.BeanContainerBuildItem;
+import io.quarkus.arc.deployment.ValidationPhaseBuildItem;
+import io.quarkus.arc.deployment.ValidationPhaseBuildItem.ValidationErrorBuildItem;
+import io.quarkus.arc.processor.BuildExtension;
+import io.quarkus.arc.processor.InjectionPointInfo;
 import io.quarkus.deployment.annotations.BuildProducer;
 import io.quarkus.deployment.annotations.BuildStep;
 import io.quarkus.deployment.annotations.ExecutionTime;
@@ -11,17 +23,23 @@ import io.quarkus.deployment.builditem.FeatureBuildItem;
 import io.quarkus.deployment.builditem.LaunchModeBuildItem;
 import io.quarkus.deployment.builditem.ShutdownContextBuildItem;
 import io.quarkus.deployment.builditem.nativeimage.ReflectiveClassBuildItem;
+import io.quarkus.mailer.MailTemplate;
 import io.quarkus.mailer.runtime.BlockingMailerImpl;
 import io.quarkus.mailer.runtime.MailClientProducer;
 import io.quarkus.mailer.runtime.MailConfig;
 import io.quarkus.mailer.runtime.MailConfigRecorder;
+import io.quarkus.mailer.runtime.MailTemplateProducer;
 import io.quarkus.mailer.runtime.MockMailboxImpl;
 import io.quarkus.mailer.runtime.ReactiveMailerImpl;
+import io.quarkus.qute.deployment.QuteProcessor;
+import io.quarkus.qute.deployment.TemplatePathBuildItem;
 import io.quarkus.runtime.RuntimeValue;
 import io.quarkus.vertx.deployment.VertxBuildItem;
 import io.vertx.ext.mail.MailClient;
 
 public class MailerProcessor {
+
+    private static final DotName MAIL_TEMPLATE = DotName.createSimple(MailTemplate.class.getName());
 
     @BuildStep
     AdditionalBeanBuildItem registerClients() {
@@ -31,7 +49,8 @@ public class MailerProcessor {
     @BuildStep
     AdditionalBeanBuildItem registerMailers() {
         return AdditionalBeanBuildItem.builder()
-                .addBeanClasses(ReactiveMailerImpl.class, BlockingMailerImpl.class, MockMailboxImpl.class)
+                .addBeanClasses(ReactiveMailerImpl.class, BlockingMailerImpl.class, MockMailboxImpl.class,
+                        MailTemplateProducer.class)
                 .build();
     }
 
@@ -67,5 +86,51 @@ public class MailerProcessor {
         recorder.configureTheMailer(beanContainer.getValue(), config, launchMode.getLaunchMode());
 
         return new MailerBuildItem(client);
+    }
+
+    @BuildStep
+    void validateMailTemplates(
+            List<TemplatePathBuildItem> templatePaths, ValidationPhaseBuildItem validationPhase,
+            BuildProducer<ValidationErrorBuildItem> validationErrors) {
+
+        Set<String> filePaths = new HashSet<String>();
+        for (TemplatePathBuildItem templatePath : templatePaths) {
+            String filePath = templatePath.getPath();
+            if (File.separatorChar != '/') {
+                filePath = filePath.replace(File.separatorChar, '/');
+            }
+            if (filePath.endsWith("html") || filePath.endsWith("htm") || filePath.endsWith("txt")) {
+                // For e-mails we only consider html and txt templates
+                filePaths.add(filePath);
+                int idx = filePath.lastIndexOf('.');
+                if (idx != -1) {
+                    // Also add version without suffix from the path
+                    // For example for "items.html" also add "items"
+                    filePaths.add(filePath.substring(0, idx));
+                }
+            }
+        }
+
+        for (InjectionPointInfo injectionPoint : validationPhase.getContext().get(BuildExtension.Key.INJECTION_POINTS)) {
+            if (injectionPoint.getRequiredType().name().equals(MAIL_TEMPLATE)) {
+                AnnotationInstance resourcePath = injectionPoint.getRequiredQualifier(QuteProcessor.RESOURCE_PATH);
+                String name;
+                if (resourcePath != null) {
+                    name = resourcePath.value().asString();
+                } else if (injectionPoint.hasDefaultedQualifier()) {
+                    name = QuteProcessor.getName(injectionPoint);
+                } else {
+                    name = null;
+                }
+                if (name != null) {
+                    // For "@Inject MailTemplate items" we try to match "items"
+                    // For "@ResourcePath("github/pulls") MailTemplate pulls" we try to match "github/pulls"
+                    if (filePaths.stream().noneMatch(path -> path.endsWith(name))) {
+                        validationErrors.produce(new ValidationErrorBuildItem(
+                                new IllegalStateException("No mail template found for " + injectionPoint.getTargetInfo())));
+                    }
+                }
+            }
+        }
     }
 }

--- a/extensions/mailer/deployment/src/test/java/io/quarkus/mailer/InjectionTest.java
+++ b/extensions/mailer/deployment/src/test/java/io/quarkus/mailer/InjectionTest.java
@@ -4,13 +4,16 @@ import java.util.concurrent.CompletionStage;
 
 import javax.enterprise.context.ApplicationScoped;
 import javax.inject.Inject;
+import javax.inject.Singleton;
 
 import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.asset.StringAsset;
 import org.jboss.shrinkwrap.api.spec.JavaArchive;
 import org.junit.jupiter.api.Assertions;
 import org.junit.jupiter.api.Test;
 import org.junit.jupiter.api.extension.RegisterExtension;
 
+import io.quarkus.qute.api.ResourcePath;
 import io.quarkus.test.QuarkusUnitTest;
 import io.vertx.ext.mail.MailClient;
 
@@ -23,7 +26,14 @@ public class InjectionTest {
             .setArchiveProducer(() -> ShrinkWrap.create(JavaArchive.class)
                     .addClasses(BeanUsingAxleMailClient.class, BeanUsingBareMailClient.class, BeanUsingRxClient.class)
                     .addClasses(BeanUsingBlockingMailer.class, BeanUsingReactiveMailer.class)
-                    .addAsResource("mock-config.properties", "application.properties"));
+                    .addClasses(MailTemplates.class)
+                    .addAsResource("mock-config.properties", "application.properties")
+                    .addAsResource(new StringAsset(""
+                            + "<html>{name}</html>"), "templates/test1.html")
+                    .addAsResource(new StringAsset(""
+                            + "{name}"), "templates/test1.txt")
+                    .addAsResource(new StringAsset(""
+                            + "<html>{name}</html>"), "templates/mails/test2.html"));
 
     @Inject
     BeanUsingAxleMailClient beanUsingBare;
@@ -40,13 +50,18 @@ public class InjectionTest {
     @Inject
     BeanUsingBlockingMailer beanUsingBlockingMailer;
 
+    @Inject
+    MailTemplates templates;
+
     @Test
     public void testInjection() {
         beanUsingAxle.verify();
         beanUsingBare.verify();
         beanUsingRx.verify();
         beanUsingBlockingMailer.verify();
-        beanUsingReactiveMailer.verify().toCompletableFuture().join();
+        beanUsingReactiveMailer.verify();
+        templates.send1();
+        templates.send2().toCompletableFuture().join();
     }
 
     @ApplicationScoped
@@ -102,5 +117,24 @@ public class InjectionTest {
         void verify() {
             mailer.send(Mail.withText("quarkus@quarkus.io", "test mailer", "blocking test!"));
         }
+    }
+
+    @Singleton
+    static class MailTemplates {
+
+        @Inject
+        MailTemplate test1;
+
+        @ResourcePath("mails/test2")
+        MailTemplate testMail;
+
+        CompletionStage<Void> send1() {
+            return test1.to("quarkus@quarkus.io").subject("Test").data("name", "John").send();
+        }
+
+        CompletionStage<Void> send2() {
+            return testMail.to("quarkus@quarkus.io").subject("Test").data("name", "Lu").send();
+        }
+
     }
 }

--- a/extensions/mailer/deployment/src/test/java/io/quarkus/mailer/MailTemplateValidationTest.java
+++ b/extensions/mailer/deployment/src/test/java/io/quarkus/mailer/MailTemplateValidationTest.java
@@ -1,0 +1,46 @@
+package io.quarkus.mailer;
+
+import java.util.concurrent.CompletionStage;
+
+import javax.enterprise.inject.spi.DeploymentException;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.asset.StringAsset;
+import org.jboss.shrinkwrap.api.spec.JavaArchive;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.test.QuarkusUnitTest;
+
+public class MailTemplateValidationTest {
+
+    @RegisterExtension
+    static final QuarkusUnitTest config = new QuarkusUnitTest()
+            .setArchiveProducer(() -> ShrinkWrap.create(JavaArchive.class)
+                    .addClasses(MailTemplates.class)
+                    .addAsResource("mock-config.properties", "application.properties")
+                    .addAsResource(new StringAsset(""
+                            + "<html>{name}</html>"), "templates/test1.html"))
+            .setExpectedException(DeploymentException.class);
+
+    @Test
+    public void testValidationFailed() {
+        // This method should not be invoked
+        Assertions.fail();
+    }
+
+    @Singleton
+    static class MailTemplates {
+
+        @Inject
+        MailTemplate doesNotExist;
+
+        CompletionStage<Void> send() {
+            return doesNotExist.to("quarkus@quarkus.io").subject("Test").data("name", "Foo").send();
+        }
+
+    }
+}

--- a/extensions/mailer/runtime/pom.xml
+++ b/extensions/mailer/runtime/pom.xml
@@ -21,10 +21,13 @@
             <artifactId>quarkus-vertx</artifactId>
         </dependency>
         <dependency>
+            <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-qute</artifactId>
+        </dependency>
+        <dependency>
             <groupId>io.smallrye.reactive</groupId>
             <artifactId>smallrye-axle-mail-client</artifactId>
         </dependency>
-
         <dependency>
             <groupId>org.subethamail</groupId>
             <artifactId>subethasmtp</artifactId>

--- a/extensions/mailer/runtime/src/main/java/io/quarkus/mailer/MailTemplate.java
+++ b/extensions/mailer/runtime/src/main/java/io/quarkus/mailer/MailTemplate.java
@@ -1,0 +1,57 @@
+package io.quarkus.mailer;
+
+import java.util.concurrent.CompletionStage;
+
+/**
+ * Represents an e-mail definition based on a template.
+ */
+public interface MailTemplate {
+
+    /**
+     * 
+     * @return a new template instance
+     */
+    MailTemplateInstance instance();
+
+    default MailTemplateInstance of(Mail mail) {
+        return instance().mail(mail);
+    }
+
+    default MailTemplateInstance to(String... values) {
+        return instance().to(values);
+    }
+
+    default MailTemplateInstance data(String key, Object value) {
+        return instance().data(key, value);
+    }
+
+    /**
+     * Represents an instance of {@link MailTemplate}.
+     * <p>
+     * This construct is not thread-safe.
+     */
+    interface MailTemplateInstance {
+
+        MailTemplateInstance mail(Mail mail);
+
+        MailTemplateInstance to(String... to);
+
+        MailTemplateInstance cc(String... cc);
+
+        MailTemplateInstance bcc(String... bcc);
+
+        MailTemplateInstance subject(String subject);
+
+        MailTemplateInstance from(String from);
+
+        MailTemplateInstance replyTo(String replyTo);
+
+        MailTemplateInstance bounceAddress(String bounceAddress);
+
+        MailTemplateInstance data(String key, Object value);
+
+        CompletionStage<Void> send();
+
+    }
+
+}

--- a/extensions/mailer/runtime/src/main/java/io/quarkus/mailer/runtime/MailTemplateInstanceImpl.java
+++ b/extensions/mailer/runtime/src/main/java/io/quarkus/mailer/runtime/MailTemplateInstanceImpl.java
@@ -1,0 +1,160 @@
+package io.quarkus.mailer.runtime;
+
+import static io.quarkus.qute.api.VariantTemplate.SELECTED_VARIANT;
+import static io.quarkus.qute.api.VariantTemplate.VARIANTS;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.ExecutionException;
+
+import io.quarkus.mailer.Mail;
+import io.quarkus.mailer.MailTemplate;
+import io.quarkus.mailer.MailTemplate.MailTemplateInstance;
+import io.quarkus.mailer.ReactiveMailer;
+import io.quarkus.qute.TemplateInstance;
+import io.quarkus.qute.Variant;
+import io.quarkus.qute.api.VariantTemplate;
+
+class MailTemplateInstanceImpl implements MailTemplate.MailTemplateInstance {
+
+    private final ReactiveMailer mailer;
+    private final TemplateInstance templateInstance;
+    private final Map<String, Object> data;
+    private Mail mail;
+
+    MailTemplateInstanceImpl(ReactiveMailer mailer, TemplateInstance templateInstance) {
+        this.mailer = mailer;
+        this.templateInstance = templateInstance;
+        this.data = new HashMap<>();
+        this.mail = new Mail();
+    }
+
+    @Override
+    public MailTemplateInstance mail(Mail mail) {
+        this.mail = mail;
+        return this;
+    }
+
+    @Override
+    public MailTemplateInstance to(String... to) {
+        this.mail.addTo(to);
+        return this;
+    }
+
+    @Override
+    public MailTemplateInstance cc(String... cc) {
+        this.mail.addCc(cc);
+        return this;
+    }
+
+    @Override
+    public MailTemplateInstance bcc(String... bcc) {
+        this.mail.addBcc(bcc);
+        return this;
+    }
+
+    @Override
+    public MailTemplateInstance subject(String subject) {
+        this.mail.setSubject(subject);
+        return this;
+    }
+
+    @Override
+    public MailTemplateInstance from(String from) {
+        this.mail.setFrom(from);
+        return this;
+    }
+
+    @Override
+    public MailTemplateInstance replyTo(String replyTo) {
+        this.mail.setReplyTo(replyTo);
+        return this;
+    }
+
+    @Override
+    public MailTemplateInstance bounceAddress(String bounceAddress) {
+        this.mail.setBounceAddress(bounceAddress);
+        return this;
+    }
+
+    @Override
+    public MailTemplateInstance data(String key, Object value) {
+        this.data.put(key, value);
+        return this;
+    }
+
+    @Override
+    public CompletionStage<Void> send() {
+
+        CompletableFuture<Void> result = new CompletableFuture<>();
+
+        if (templateInstance.getAttribute(VariantTemplate.VARIANTS) != null) {
+
+            List<Result> results = new ArrayList<>();
+
+            @SuppressWarnings("unchecked")
+            List<Variant> variants = (List<Variant>) templateInstance.getAttribute(VARIANTS);
+            for (Variant variant : variants) {
+                if (variant.mediaType.equals(Variant.TEXT_HTML) || variant.mediaType.equals(Variant.TEXT_PLAIN)) {
+                    results.add(new Result(variant,
+                            templateInstance.setAttribute(SELECTED_VARIANT, variant).data(data).renderAsync()
+                                    .toCompletableFuture()));
+                }
+            }
+
+            if (results.isEmpty()) {
+                throw new IllegalStateException("No suitable template variant found");
+            }
+
+            CompletableFuture<Void> all = CompletableFuture
+                    .allOf(results.stream().map(Result::getValue).toArray(CompletableFuture[]::new));
+            all.whenComplete((r1, t1) -> {
+                if (t1 != null) {
+                    result.completeExceptionally(t1);
+                } else {
+                    try {
+                        for (Result res : results) {
+                            if (res.variant.mediaType.equals(Variant.TEXT_HTML)) {
+                                mail.setHtml(res.value.get());
+                            } else if (res.variant.mediaType.equals(Variant.TEXT_PLAIN)) {
+                                mail.setText(res.value.get());
+                            }
+                        }
+                    } catch (InterruptedException | ExecutionException e) {
+                        result.completeExceptionally(e);
+                    }
+                    mailer.send(mail).whenComplete((r, t) -> {
+                        if (t != null) {
+                            result.completeExceptionally(t);
+                        } else {
+                            result.complete(null);
+                        }
+                    });
+                }
+            });
+        } else {
+            throw new IllegalStateException("No template variant found");
+        }
+        return result;
+    }
+
+    static class Result {
+
+        final Variant variant;
+        final CompletableFuture<String> value;
+
+        public Result(Variant variant, CompletableFuture<String> result) {
+            this.variant = variant;
+            this.value = result;
+        }
+
+        CompletableFuture<String> getValue() {
+            return value;
+        }
+    }
+
+}

--- a/extensions/mailer/runtime/src/main/java/io/quarkus/mailer/runtime/MailTemplateProducer.java
+++ b/extensions/mailer/runtime/src/main/java/io/quarkus/mailer/runtime/MailTemplateProducer.java
@@ -1,0 +1,79 @@
+package io.quarkus.mailer.runtime;
+
+import java.lang.annotation.Annotation;
+import java.lang.reflect.Field;
+
+import javax.enterprise.inject.Any;
+import javax.enterprise.inject.Instance;
+import javax.enterprise.inject.Produces;
+import javax.enterprise.inject.spi.AnnotatedParameter;
+import javax.enterprise.inject.spi.InjectionPoint;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+
+import org.jboss.logging.Logger;
+
+import io.quarkus.mailer.MailTemplate;
+import io.quarkus.mailer.ReactiveMailer;
+import io.quarkus.qute.api.ResourcePath;
+import io.quarkus.qute.api.VariantTemplate;
+
+@Singleton
+public class MailTemplateProducer {
+
+    private static final Logger LOGGER = Logger.getLogger(MailTemplateProducer.class);
+
+    @Inject
+    ReactiveMailer mailer;
+
+    @Any
+    Instance<VariantTemplate> template;
+
+    @Produces
+    MailTemplate getDefault(InjectionPoint injectionPoint) {
+
+        final String name;
+        if (injectionPoint.getMember() instanceof Field) {
+            // For "@Inject MailTemplate test" use "test"
+            name = injectionPoint.getMember().getName();
+        } else {
+            AnnotatedParameter<?> parameter = (AnnotatedParameter<?>) injectionPoint.getAnnotated();
+            if (parameter.getJavaParameter().isNamePresent()) {
+                name = parameter.getJavaParameter().getName();
+            } else {
+                name = injectionPoint.getMember().getName();
+                LOGGER.warnf("Parameter name not present - using the method name as the template name instead %s", name);
+            }
+        }
+
+        return new MailTemplate() {
+            @Override
+            public MailTemplateInstance instance() {
+                return new MailTemplateInstanceImpl(mailer, template.select(new ResourcePath.Literal(name)).get().instance());
+            }
+
+        };
+    }
+
+    @ResourcePath("ignored")
+    @Produces
+    MailTemplate get(InjectionPoint injectionPoint) {
+        ResourcePath path = null;
+        for (Annotation qualifier : injectionPoint.getQualifiers()) {
+            if (qualifier.annotationType().equals(ResourcePath.class)) {
+                path = (ResourcePath) qualifier;
+                break;
+            }
+        }
+        if (path == null || path.value().isEmpty()) {
+            throw new IllegalStateException("No template reource path specified");
+        }
+        final String name = path.value();
+        return new MailTemplate() {
+            @Override
+            public MailTemplateInstance instance() {
+                return new MailTemplateInstanceImpl(mailer, template.select(new ResourcePath.Literal(name)).get().instance());
+            }
+        };
+    }
+}

--- a/extensions/qute/deployment/src/main/java/io/quarkus/qute/deployment/QuteProcessor.java
+++ b/extensions/qute/deployment/src/main/java/io/quarkus/qute/deployment/QuteProcessor.java
@@ -275,7 +275,7 @@ public class QuteProcessor {
     }
 
     @BuildStep
-    void validateInjectedBeans(ApplicationArchivesBuildItem applicationArchivesBuildItem,
+    void validateBeansInjectedInTemplates(ApplicationArchivesBuildItem applicationArchivesBuildItem,
             TemplatesAnalysisBuildItem analysis, BeanArchiveIndexBuildItem beanArchiveIndex,
             BuildProducer<IncorrectExpressionBuildItem> incorrectExpressions,
             List<TemplateExtensionMethodBuildItem> templateExtensionMethods,
@@ -496,30 +496,20 @@ public class QuteProcessor {
     }
 
     @BuildStep
-    void processInjectionPoints(QuteConfig config, ApplicationArchivesBuildItem applicationArchivesBuildItem,
-            List<TemplatePathBuildItem> templatePaths, ValidationPhaseBuildItem validationPhase,
-            BuildProducer<ValidationErrorBuildItem> validationErrors, BuildProducer<TemplateVariantsBuildItem> templateVariants)
-            throws IOException {
+    void validateTemplateInjectionPoints(List<TemplatePathBuildItem> templatePaths, ValidationPhaseBuildItem validationPhase,
+            BuildProducer<ValidationErrorBuildItem> validationErrors) {
 
-        ApplicationArchive applicationArchive = applicationArchivesBuildItem.getRootArchive();
-        String basePath = "templates/";
-        Path templatesPath = applicationArchive.getChildPath(basePath);
-
-        // Remove suffix from the path; e.g. "items.html" becomes "items"
         Set<String> filePaths = new HashSet<String>();
         for (TemplatePathBuildItem templatePath : templatePaths) {
-            String filePath = templatePath.getPath();
-            if (File.separatorChar != '/') {
-                filePath = filePath.replace(File.separatorChar, '/');
-            }
-            filePaths.add(filePath);
-            int idx = filePath.lastIndexOf('.');
+            String path = templatePath.getPath();
+            filePaths.add(path);
+            int idx = path.lastIndexOf('.');
             if (idx != -1) {
-                filePaths.add(filePath.substring(0, idx));
+                // Also add version without suffix from the path
+                // For example for "items.html" also add "items"
+                filePaths.add(path.substring(0, idx));
             }
         }
-
-        Set<String> variantBases = new HashSet<>();
 
         for (InjectionPointInfo injectionPoint : validationPhase.getContext().get(BuildExtension.Key.INJECTION_POINTS)) {
 
@@ -558,19 +548,31 @@ public class QuteProcessor {
                     if (filePaths.stream().noneMatch(path -> path.endsWith(name))) {
                         validationErrors.produce(new ValidationErrorBuildItem(
                                 new IllegalStateException("No variant template found for " + injectionPoint.getTargetInfo())));
-                    } else {
-                        variantBases.add(name);
                     }
                 }
             }
         }
+    }
 
-        if (!variantBases.isEmpty()) {
-            Map<String, List<String>> variants = new HashMap<>();
-            scanVariants(basePath, templatesPath, templatesPath, variantBases, variants);
-            templateVariants.produce(new TemplateVariantsBuildItem(variants));
-            LOGGER.debugf("Variant templates found: %s", variants);
+    @BuildStep
+    TemplateVariantsBuildItem collectTemplateVariants(List<TemplatePathBuildItem> templatePaths) throws IOException {
+        Set<String> allPaths = templatePaths.stream().map(TemplatePathBuildItem::getPath).collect(Collectors.toSet());
+        // item -> [item.html, item.txt]
+        Map<String, List<String>> baseToVariants = new HashMap<>();
+        for (String path : allPaths) {
+            int idx = path.lastIndexOf('.');
+            if (idx != -1) {
+                String base = path.substring(0, idx);
+                List<String> variants = baseToVariants.get(base);
+                if (variants == null) {
+                    variants = new ArrayList<>();
+                    baseToVariants.put(base, variants);
+                }
+                variants.add(path);
+            }
         }
+        LOGGER.debugf("Variant templates found: %s", baseToVariants);
+        return new TemplateVariantsBuildItem(baseToVariants);
     }
 
     @BuildStep
@@ -781,7 +783,7 @@ public class QuteProcessor {
         return injectExpressions;
     }
 
-    private String getName(InjectionPointInfo injectionPoint) {
+    public static String getName(InjectionPointInfo injectionPoint) {
         if (injectionPoint.isField()) {
             return injectionPoint.getTarget().asField().name();
         } else if (injectionPoint.isParam()) {
@@ -813,36 +815,19 @@ public class QuteProcessor {
         try (Stream<Path> files = Files.list(directory)) {
             Iterator<Path> iter = files.iterator();
             while (iter.hasNext()) {
-                Path path = iter.next();
-                if (Files.isRegularFile(path)) {
-                    LOGGER.debugf("Found template: %s", path);
-                    String templatePath = root.relativize(path).toString();
-                    produceTemplateBuildItems(templatePaths, watchedPaths, nativeImageResources, basePath, templatePath, path,
-                            false);
-                } else if (Files.isDirectory(path) && !path.getFileName().toString().equals("tags")) {
-                    LOGGER.debugf("Scan directory: %s", path);
-                    scan(root, path, basePath, watchedPaths, templatePaths, nativeImageResources);
-                }
-            }
-        }
-    }
-
-    void scanVariants(String basePath, Path root, Path directory, Set<String> variantBases, Map<String, List<String>> variants)
-            throws IOException {
-        try (Stream<Path> files = Files.list(directory)) {
-            Iterator<Path> iter = files.iterator();
-            while (iter.hasNext()) {
-                Path path = iter.next();
-                if (Files.isRegularFile(path)) {
-                    for (String base : variantBases) {
-                        if (path.toAbsolutePath().toString().contains(base)) {
-                            // Variants are relative paths to base, e.g. "detail/item2"
-                            variants.computeIfAbsent(base, i -> new ArrayList<>())
-                                    .add(root.relativize(path).toString());
-                        }
+                Path filePath = iter.next();
+                if (Files.isRegularFile(filePath)) {
+                    LOGGER.debugf("Found template: %s", filePath);
+                    String templatePath = root.relativize(filePath).toString();
+                    if (File.separatorChar != '/') {
+                        templatePath = templatePath.replace(File.separatorChar, '/');
                     }
-                } else if (Files.isDirectory(path)) {
-                    scanVariants(basePath, root, path, variantBases, variants);
+                    produceTemplateBuildItems(templatePaths, watchedPaths, nativeImageResources, basePath, templatePath,
+                            filePath,
+                            false);
+                } else if (Files.isDirectory(filePath) && !filePath.getFileName().toString().equals("tags")) {
+                    LOGGER.debugf("Scan directory: %s", filePath);
+                    scan(root, filePath, basePath, watchedPaths, templatePaths, nativeImageResources);
                 }
             }
         }

--- a/extensions/qute/deployment/src/main/java/io/quarkus/qute/deployment/TemplatePathBuildItem.java
+++ b/extensions/qute/deployment/src/main/java/io/quarkus/qute/deployment/TemplatePathBuildItem.java
@@ -13,24 +13,34 @@ public final class TemplatePathBuildItem extends MultiBuildItem {
     private final Path fullPath;
     private final boolean tag;
 
-    public TemplatePathBuildItem(String path, Path fullPath) {
-        this(path, fullPath, false);
-    }
-
     public TemplatePathBuildItem(String path, Path fullPath, boolean tag) {
         this.path = path;
         this.fullPath = fullPath;
         this.tag = tag;
     }
 
+    /**
+     * Uses the {@code /} path separator.
+     * 
+     * @return the path relative to the template root
+     */
     public String getPath() {
         return path;
     }
 
+    /**
+     * Uses the system-dependent path separator.
+     * 
+     * @return the full path of the template
+     */
     public Path getFullPath() {
         return fullPath;
     }
 
+    /**
+     * 
+     * @return {@code true} if it represents a user tag, {@code false} otherwise
+     */
     public boolean isTag() {
         return tag;
     }

--- a/extensions/qute/runtime/src/main/java/io/quarkus/qute/api/ResourcePath.java
+++ b/extensions/qute/runtime/src/main/java/io/quarkus/qute/api/ResourcePath.java
@@ -8,6 +8,7 @@ import static java.lang.annotation.RetentionPolicy.RUNTIME;
 import java.lang.annotation.Retention;
 import java.lang.annotation.Target;
 
+import javax.enterprise.util.AnnotationLiteral;
 import javax.enterprise.util.Nonbinding;
 import javax.inject.Qualifier;
 
@@ -25,5 +26,25 @@ public @interface ResourcePath {
      */
     @Nonbinding
     String value();
+
+    /**
+     * Supports inline instantiation of this qualifier.
+     */
+    public static final class Literal extends AnnotationLiteral<ResourcePath> implements ResourcePath {
+
+        private static final long serialVersionUID = 1L;
+
+        private final String value;
+
+        public Literal(String value) {
+            this.value = value;
+        }
+
+        @Override
+        public String value() {
+            return value;
+        }
+
+    }
 
 }

--- a/extensions/qute/runtime/src/main/java/io/quarkus/qute/runtime/QuteConfig.java
+++ b/extensions/qute/runtime/src/main/java/io/quarkus/qute/runtime/QuteConfig.java
@@ -12,8 +12,7 @@ public class QuteConfig {
     /**
      * The set of suffixes used when attempting to locate a template file.
      * 
-     * By default, `engine.getTemplate("foo")` would result in several lookups: `src/main/resources/templates/foo`,
-     * `src/main/resources/templates/foo.html` and `src/main/resources/templates/foo.txt`.
+     * By default, `engine.getTemplate("foo")` would result in several lookups: `foo`, `foo.html`, `foo.txt`, etc.
      * 
      * @asciidoclet
      */

--- a/extensions/qute/runtime/src/main/java/io/quarkus/qute/runtime/VariantTemplateProducer.java
+++ b/extensions/qute/runtime/src/main/java/io/quarkus/qute/runtime/VariantTemplateProducer.java
@@ -7,6 +7,7 @@ import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
 import java.util.Map.Entry;
+import java.util.Objects;
 import java.util.Optional;
 import java.util.Set;
 import java.util.concurrent.CompletionStage;
@@ -127,7 +128,7 @@ public class VariantTemplateProducer {
         private final TemplateVariants variants;
 
         VariantTemplateInstanceImpl(TemplateVariants variants) {
-            this.variants = variants;
+            this.variants = Objects.requireNonNull(variants);
             setAttribute(VariantTemplate.VARIANTS, new ArrayList<>(variants.variantToTemplate.keySet()));
         }
 

